doc/man/man3/Highlighter.3
.TH "Highlighter" 3 "Fri Jun 12 2015" "Version 0.0.1" "RIDE" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Highlighter \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Highlighter\&.h>\fP
.PP
Inherits QSyntaxHighlighter\&.
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBHighlightingRule\fP"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBHighlighter\fP (QTextDocument *parent=0)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBhighlightBlock\fP (const QString &text)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "QVector< \fBHighlightingRule\fP > \fBhighlightingRules\fP"
.br
.ti -1c
.RI "QRegExp \fBcommentStartExpression\fP"
.br
.ti -1c
.RI "QRegExp \fBcommentEndExpression\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBkeywordFormat\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBclassFormat\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBsingleLineCommentFormat\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBmultiLineCommentFormat\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBquotationFormat\fP"
.br
.ti -1c
.RI "QTextCharFormat \fBfunctionFormat\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 22 of file Highlighter\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Highlighter::Highlighter (QTextDocument *parent = \fC0\fP)"

.PP
Definition at line 3 of file Highlighter\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "void Highlighter::highlightBlock (const QString &text)\fC [protected]\fP"

.PP
Definition at line 158 of file Highlighter\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "QTextCharFormat Highlighter::classFormat\fC [private]\fP"

.PP
Definition at line 38 of file Highlighter\&.h\&.
.SS "QRegExp Highlighter::commentEndExpression\fC [private]\fP"

.PP
Definition at line 35 of file Highlighter\&.h\&.
.SS "QRegExp Highlighter::commentStartExpression\fC [private]\fP"

.PP
Definition at line 34 of file Highlighter\&.h\&.
.SS "QTextCharFormat Highlighter::functionFormat\fC [private]\fP"

.PP
Definition at line 42 of file Highlighter\&.h\&.
.SS "QVector<\fBHighlightingRule\fP> Highlighter::highlightingRules\fC [private]\fP"

.PP
Definition at line 32 of file Highlighter\&.h\&.
.SS "QTextCharFormat Highlighter::keywordFormat\fC [private]\fP"

.PP
Definition at line 37 of file Highlighter\&.h\&.
.SS "QTextCharFormat Highlighter::multiLineCommentFormat\fC [private]\fP"

.PP
Definition at line 40 of file Highlighter\&.h\&.
.SS "QTextCharFormat Highlighter::quotationFormat\fC [private]\fP"

.PP
Definition at line 41 of file Highlighter\&.h\&.
.SS "QTextCharFormat Highlighter::singleLineCommentFormat\fC [private]\fP"

.PP
Definition at line 39 of file Highlighter\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for RIDE from the source code\&.


src/Highlighter.cpp
#include "Highlighter.h"

Highlighter::Highlighter(QTextDocument *parent) : QSyntaxHighlighter(parent)
{
    HighlightingRule rule;

    
    keywordFormat.setForeground(Qt::blue);
    //keywordFormat.setFontWeight(QFont::Bold);
    
    //http://en.cppreference.com/w/cpp/keyword
    QStringList cPlusPlusKeywordPatterns;
    cPlusPlusKeywordPatterns << "alignas"      << "else"            << "requires"
                             << "alignof"      << "enum"             << "return"
                             << "and"          << "explicit"         << "short"
                             << "and_eq"       << "export"           << "signed"
                             << "asm"          << "extern"           << "sizeof"
                             << "auto"         << "false"            << "static"
                             << "bitand"       << "float"            << "static_assert"
                             << "bitor"        << "for"              << "static_cast"
                             << "bool"         << "friend"           << "struct"
                             << "break"        << "goto"             << "switch"
                             << "case"         << "if"               << "template"
                             << "catch"        << "inline"           << "this"
                             << "char"         << "int"              << "thread_local"
                             << "char16_t"     << "long"             << "throw"
                             << "char32_t"     << "mutable"          << "true"
                             << "class"        << "namespace"        << "try"
                             << "compl"        << "new"              << "typedef"
                             << "concept"      << "noexcept"         << "typeid"
                             << "const"        << "not"              << "typename"
                             << "constexpr"    << "not_eq"           << "union"
                             << "const_cast"   << "nullptr"          << "unsigned"
                             << "continue"     << "operator"         << "using"
                             << "decltype"     << "or"               << "virtual"
                             << "default"      << "or_eq"            << "void"
                             << "delete"       << "private"          << "volatile"
                             << "do"           << "protected"        << "wchar_t"
                             << "double"       << "public"           << "while"
                             << "dynamic_cast" << "register"         << "xor"
                                                     << "reinterpret_cast" << "xor_eq";
    QStringList cPlusPlusIdentifiers;
    cPlusPlusIdentifiers     << "override"    << "final";
    
    QStringList cPlusPlusPreprocessorCmds;
    cPlusPlusPreprocessorCmds << "#elif"         << "#ifdef"            << "#include"
                              << "#endif"        << "#ifndef"           << "#line"
                              << "#defined"      << "#define"           << "#error"
                                                      << "#undef"            << "#pragma";
    
    QStringList javaKeywordPatterns;
    javaKeywordPatterns      << "abstract"      << "assert"           << "boolean"
                             << "break"         << "byte"             << "case"
                             << "catch"         << "char"             << "class"
                             << "const"         << "continue"         << "default"
                             << "do"            << "double"           << "else"
                             << "enum"          << "extends"          << "final" 
                             << "finally"       << "float"            << "for"
                             << "goto"          << "if"               << "implements"
                             << "import"        << "instanceof"       << "int"
                             << "interface"     << "long"             << "native" 
                             << "new"           << "package"          << "private"
                             << "protected"     << "public"           << "return"
                             << "short"         << "static"           << "strictfp"
                             << "super"         << "switch"           << "synchronized"
                             << "this"          << "throw"            << "throws"
                             << "transient"     << "try"              << "void"
                             << "volatile"      << "while";
            
    QStringList pythonKeywordPatterns;
    pythonKeywordPatterns    << "False"         << "None"             << "True"
                             << "and"           << "as"               << "assert"
                             << "break"         << "class"            << "continue"
                             << "def"           << "del"              << "elif"
                             << "else"          << "except"           << "finally"
                             << "for"           << "from"             << "global"
                             << "if"            << "import"           << "in"
                             << "is"            << "lambda"           << "nonlocal"
                             << "not"           << "or"               << "pass"
                             << "raise"         << "return"           << "try"
                             << "while"         << "with"             << "yield";
    
    
    QStringList mirrorKeyCharPatterns;
    mirrorKeyCharPatterns    << "\\("             << "\\{"                << "\\["
                             << "\\)"             << "\\}"                << "\\]";
            
            
    foreach (const QString &pattern, cPlusPlusKeywordPatterns)
    {
        rule.pattern = QRegExp(pattern);
        rule.format = keywordFormat;
        highlightingRules.append(rule);
    }
    
    keywordFormat.setForeground(Qt::blue);
    //keywordFormat.setFontWeight(QFont::Bold);
    foreach (const QString &pattern, cPlusPlusIdentifiers)
    {
        rule.pattern = QRegExp(pattern);
        rule.format = keywordFormat;
        highlightingRules.append(rule);
    }
    
    QColor pur(200, 200, 100);
    keywordFormat.setForeground(pur);
    //keywordFormat.setFontWeight(QFont::Bold);
    foreach (const QString &pattern, cPlusPlusPreprocessorCmds)
    {
        rule.pattern = QRegExp(pattern);
        rule.format = keywordFormat;
        highlightingRules.append(rule);
    }
    
    keywordFormat.setForeground(Qt::red);
    //keywordFormat.setBackground(Qt::darkYellow);
    //keywordFormat.setFontWeight(QFont::Bold);
    foreach (const QString &pattern, mirrorKeyCharPatterns)
    {
        rule.pattern = QRegExp(pattern);
        rule.format = keywordFormat;
        highlightingRules.append(rule);
    }
    

    //classFormat.setFontWeight(QFont::Bold);
    classFormat.setForeground(Qt::darkMagenta);
    
    //find QT specific types (or at least type beginning with 'Q').
    rule.pattern = QRegExp("Q[A-Za-z]+");
    rule.format = classFormat;
    highlightingRules.append(rule);

    singleLineCommentFormat.setForeground(Qt::blue);
    rule.pattern = QRegExp("//[^\n]*");
    rule.format = singleLineCommentFormat;
    highlightingRules.append(rule);

    multiLineCommentFormat.setForeground(Qt::blue);

    quotationFormat.setForeground(Qt::red);
    rule.pattern = QRegExp("\".*\"");
    rule.format = quotationFormat;
    highlightingRules.append(rule);

    //functionFormat.setFontItalic(true);
    functionFormat.setFontWeight(QFont::Bold);
    //QColor purple(147, 112, 219);
    //functionFormat.setForeground(purple);
    rule.pattern = QRegExp("[A-Za-z0-9_]+(?=\\()");
    rule.format = functionFormat;
    highlightingRules.append(rule);

    commentStartExpression = QRegExp("/\\*");
    commentEndExpression = QRegExp("\\*/");
}

void Highlighter::highlightBlock(const QString &text)
{
    foreach (const HighlightingRule &rule, highlightingRules)
    {
        QRegExp expression(rule.pattern);
        int index = expression.indexIn(text);
        while (index >= 0)
        {
            int length = expression.matchedLength();
            setFormat(index, length, rule.format);
            index = expression.indexIn(text, index + length);
        }
    }
    setCurrentBlockState(0);

    int startIndex = 0;
    if(previousBlockState() != 1)
        startIndex = commentStartExpression.indexIn(text);

    while (startIndex >= 0)
    {
        int endIndex = commentEndExpression.indexIn(text, startIndex);
        int commentLength;
        if (endIndex == -1)
        {
            setCurrentBlockState(1);
            commentLength = text.length() - startIndex;
        }
        else
        {
            commentLength = endIndex - startIndex
                            + commentEndExpression.matchedLength();
        }
        setFormat(startIndex, commentLength, multiLineCommentFormat);
        startIndex = commentStartExpression.indexIn(text, startIndex + commentLength);
    }
}


QString* Highlighter::toString()
{
    QString* tmp = new QString();
    tmp->append("***method stub***");
    
    return tmp;
}


Highlighter::~Highlighter()
{
    ;
}